src/Voron/Data/CompactTrees/PersistentDictionary.cs

```diff
@@ -0,0 +1,199 @@
+        }
+
+        public static PersistentDictionary Create<TKeysEnumerator>(LowLevelTransaction llt, in TKeysEnumerator enumerator, PersistentDictionary previousDictionary = null)
+            where TKeysEnumerator : struct, IReadOnlySpanEnumerator
+        {
+            var p = llt.AllocatePage(NumberOfPagesForDictionary);
+            p.Flags = PageFlags.Overflow;
+            p.OverflowSize = NumberOfPagesForDictionary * Constants.Storage.PageSize;
+
+            PersistentDictionaryHeader* header = (PersistentDictionaryHeader*)p.DataPointer;
+            header->CurrentId = p.PageNumber;
+            header->PreviousId = previousDictionary != null ? previousDictionary.PageNumber : 0;
+            
+            var encoder = new HopeEncoder<Encoder3Gram<NativeMemoryEncoderState>>(
+                new Encoder3Gram<NativeMemoryEncoderState>(
+                    new NativeMemoryEncoderState(p.DataPointer + sizeof(PersistentDictionaryHeader), UsableDictionarySize)));
+            encoder.Train(enumerator, MaxDictionarySize);
+
+            header->TableHash = XXHash64.Calculate(p.DataPointer + sizeof(long), UsableDictionarySize + PersistentDictionaryHeader.SizeOf - sizeof(long));
+
+#if DEBUG
+            VerifyTable(p);
+#endif
+
+            return new PersistentDictionary(p);
+        }
+
+
+        public static PersistentDictionary CreateIfBetter<TKeysEnumerator>(LowLevelTransaction llt, in TKeysEnumerator trainEnumerator, in TKeysEnumerator testEnumerator, PersistentDictionary previousDictionary = null)
+            where TKeysEnumerator : struct, IReadOnlySpanEnumerator
+        {
+            // Reserve the memory to create the encoder table. 
+            using var scope = llt.Allocator.Allocate(UsableDictionarySize, out var output);
+            
+            // Train the new dictionary.
+            var encoder = new HopeEncoder<Encoder3Gram<NativeMemoryEncoderState>>(
+                                new Encoder3Gram<NativeMemoryEncoderState>(
+                                    new NativeMemoryEncoderState(output.Ptr, UsableDictionarySize)));
+            encoder.Train(trainEnumerator, MaxDictionarySize);
+
+            // Test the new dictionary to ensure that we have statistically better compression.
+            using var encodeBufferScope = llt.Allocator.Allocate(Constants.Storage.PageSize, out var encodeBuffer);
+            
+            int incumbentSize = 0;
+            int successorSize = 0;
+            var auxEncodeBuffer = encodeBuffer.ToSpan();
+            while (testEnumerator.MoveNext(out var testValue))
+            {                
+                incumbentSize += previousDictionary._encoder.Encode(testValue, auxEncodeBuffer);
+                successorSize += encoder.Encode(testValue, auxEncodeBuffer);
+            }
+
+            // If the new dictionary is not at least 2.5% better, we return the current dictionary.             
+            if (incumbentSize < successorSize * 1.025)
```

Comment:
That seems like a pretty small benefit, no? 
I would say 15% or 25% is the threshold to want to move to a new one

Reply:
If we have a bigger threshold, we can also then set regular retraining intervals

Reply:
In machine learning I use a rule of thumb way, I switch models if 5% better. There are stability and statistical goodies on why to do so for competing models. I decided to use 2.5% instead (not convinced though) because at some sizes (thinking 10Gb or so, even 2% is big [200Mb]). 5% is a number I do like though. At some point you converge and no matter what you do, you wont get more than a 1%, and at that time; it becomes really worthless to even try anymore. You are optimal. If we ever hit that place, we should be thinking into increasing the size of dictionary.